src/vaults/PoolManager.sol

```diff
@@ -287,27 +301,48 @@ contract PoolManager is Auth, IPoolManager, IUpdateContract, IPoolManagerGateway
     /// @notice The pool manager either deploys the vault if a factory address is provided or it simply links/unlinks
     /// the vault
     function update(uint64 poolId, bytes16 scId, bytes memory payload) public auth {
-        MessageLib.UpdateContractVaultUpdate memory m = MessageLib.deserializeUpdateContractVaultUpdate(payload);
-
-        if (m.kind == uint8(VaultUpdateKind.DeployAndLink)) {
-            address factory = address(bytes20(m.vaultOrFactory));
+        uint8 kind = uint8(MessageLib.updateContractType(payload));
 
-            address vault = deployVault(poolId, scId, m.assetId, factory);
-            linkVault(poolId, scId, m.assetId, vault);
-        } else {
-            address vault = address(bytes20(m.vaultOrFactory));
+        if (kind == uint8(UpdateContractType.VaultUpdate)) {
+            MessageLib.UpdateContractVaultUpdate memory m = MessageLib.deserializeUpdateContractVaultUpdate(payload);
 
-            // Needed as safeguard against non-validated vaults
-            // I.e. we only accept vaults that have been deployed by the pool manager
-            require(_vaultDetails[vault].asset != address(0), "PoolManager/unknown-vault");
+            if (m.kind == uint8(VaultUpdateKind.DeployAndLink)) {
+                address factory = address(bytes20(m.vaultOrFactory));
 
-            if (m.kind == uint8(VaultUpdateKind.Link)) {
+                address vault = deployVault(poolId, scId, m.assetId, factory);
                 linkVault(poolId, scId, m.assetId, vault);
-            } else if (m.kind == uint8(VaultUpdateKind.Unlink)) {
-                unlinkVault(poolId, scId, m.assetId, vault);
             } else {
-                revert("PoolManager/malformed-vault-update-msg");
+                address vault = address(bytes20(m.vaultOrFactory));
+
+                // Needed as safeguard against non-validated vaults
+                // I.e. we only accept vaults that have been deployed by the pool manager
+                require(_vaultDetails[vault].asset != address(0), "PoolManager/unknown-vault");
+
+                if (m.kind == uint8(VaultUpdateKind.Link)) {
+                    linkVault(poolId, scId, m.assetId, vault);
+                } else if (m.kind == uint8(VaultUpdateKind.Unlink)) {
+                    unlinkVault(poolId, scId, m.assetId, vault);
+                } else {
+                    revert("PoolManager/malformed-vault-update-msg");
+                }
             }
+        } else if (kind == uint8(UpdateContractType.MaxPriceAge)) {
+            MessageLib.UpdateContractMaxPriceAge memory m = MessageLib.deserializeUpdateContractMaxPriceAge(payload);
+
+            ShareClassDetails storage shareClass = pools[poolId].shareClasses[scId];
+            require(shareClass.shareToken != address(0), "PoolManager/share-token-does-not-exist");
+
+            // Unset asset id acts as wildcard for setting share max price age
+            if (m.assetId == 0) {
```

Comment:
Thanks for the comment, that was weird! Could this deserve its own message?

Reply:
I split it up now